controller/query.go

```diff
@@ -0,0 +1,147 @@
+package controller
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/fabric8-services/fabric8-wit/app"
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/jsonapi"
+	"github.com/fabric8-services/fabric8-wit/login"
+	"github.com/fabric8-services/fabric8-wit/query"
+	"github.com/fabric8-services/fabric8-wit/rest"
+	"github.com/fabric8-services/fabric8-wit/search"
+	"github.com/goadesign/goa"
+	"github.com/prometheus/common/log"
+)
+
+// QueryController implements the query resource.
+type QueryController struct {
+	*goa.Controller
+	db     application.DB
+	config QueryControllerConfiguration
+}
+
+// QueryControllerConfiguration the configuration for the LabelController
+type QueryControllerConfiguration interface {
+	GetCacheControlQueries() string
+	GetCacheControlQuery() string
+}
+
+// NewQueryController creates a query controller.
+func NewQueryController(service *goa.Service, db application.DB, config QueryControllerConfiguration) *QueryController {
+	return &QueryController{
+		Controller: service.NewController("QueryController"),
+		db:         db,
+		config:     config,
+	}
+}
+
+// Create runs the create action.
+func (c *QueryController) Create(ctx *app.CreateQueryContext) error {
+	currentUserIdentityID, err := login.ContextIdentity(ctx)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		q := query.Query{
+			SpaceID: ctx.SpaceID,
+			Fields:  ctx.Payload.Data.Attributes.Fields,
+			Title:   strings.TrimSpace(ctx.Payload.Data.Attributes.Title),
+			Creator: *currentUserIdentityID,
+		}
+		// Parse fields to make sure that query is valid
+		_, _, err := search.ParseFilterString(ctx, q.Fields)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"space_id": ctx.SpaceID,
+				"fields":   q.Fields,
+			}, "unable to parse the query fields")
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		err = appl.Queries().Create(ctx, &q)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		res := &app.QuerySingle{
+			Data: ConvertQuery(appl, ctx.Request, q),
+		}
+		ctx.ResponseData.Header().Set("Location", rest.AbsoluteURL(ctx.Request, app.LabelHref(ctx.SpaceID, res.Data.ID)))
+		return ctx.Created(res)
+	})
+}
+
+// ConvertQuery converts from internal to external REST representation
+func ConvertQuery(appl application.Application, request *http.Request, q query.Query) *app.Query {
+	queryType := query.APIStringTypeQuery
+	spaceID := q.SpaceID.String()
+	relatedURL := rest.AbsoluteURL(request, app.LabelHref(spaceID, q.ID))
```

Comment:
The `LabelHref` should be replaced with `QueryHref`

Reply:
corrected in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4